temporal-sdk/src/main/java/io/temporal/internal/common/SearchAttributePayloadConverter.java

```diff
@@ -0,0 +1,254 @@
+/*
+ *  Copyright (C) 2020 Temporal Technologies, Inc. All Rights Reserved.
+ *
+ *  Copyright 2012-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *  Modifications copyright (C) 2017 Uber Technologies, Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"). You may not
+ *  use this file except in compliance with the License. A copy of the License is
+ *  located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ *  or in the "license" file accompanying this file. This file is distributed on
+ *  an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *  express or implied. See the License for the specific language governing
+ *  permissions and limitations under the License.
+ */
+
+package io.temporal.internal.common;
+
+import com.google.common.base.Preconditions;
+import com.google.common.reflect.TypeParameter;
+import com.google.common.reflect.TypeToken;
+import com.google.protobuf.ByteString;
+import io.temporal.api.common.v1.Payload;
+import io.temporal.api.enums.v1.IndexedValueType;
+import io.temporal.common.SearchAttribute;
+import io.temporal.common.converter.DataConverterException;
+import io.temporal.common.converter.DefaultDataConverter;
+import java.lang.reflect.Type;
+import java.time.OffsetDateTime;
+import java.util.*;
+import java.util.stream.Collectors;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+final class SearchAttributePayloadConverter {
+  private static final Logger log = LoggerFactory.getLogger(SearchAttributePayloadConverter.class);
+
+  private static final String METADATA_TYPE_KEY = "type";
+
+  private final DefaultDataConverter defaultDataConverter =
+      DefaultDataConverter.newDefaultInstance();
+  public static final SearchAttributePayloadConverter INSTANCE =
+      new SearchAttributePayloadConverter();
+
+  /**
+   * Convert Search Attribute object into payload with metadata. Ideally, we don't want to send the
+   * type metadata to the server, because starting with v1.10.0 Temporal doesn't look at the type
+   * metadata that the SDK sends. When the attribute is registered with the service, so is its
+   * intended type which is used to validate the data. However, we do include type metadata in
+   * payload here for compatibility with older versions of the server. Earlier version of Temporal
+   * save the type metadata and return exactly the same payload back to the SDK, which will be
+   * needed to deserialize the attribute into it's initial type.
+   */
+  public Payload encode(@Nonnull Object obj) {
+    if (obj instanceof Collection && ((Collection<?>) obj).size() == 1) {
+      // always serialize an empty collection as one value
+      obj = ((Collection<?>) obj).iterator().next();
```

Comment:
Hrmm, it is a bit confusing that if I encode a string list of size 1 and decode it I get a string but if I encode a string list of size 2 and decode it I get a list. If we did encode this as a single-item list, would the server just send us back a single item anyways?